apps/solution/src/components/EditPerson/UseManagers.tsx

```diff
@@ -0,0 +1,14 @@
+import { useEffect, useMemo, useState } from 'react';
+import { getPeople, PersonModel } from '../../api/person';
+
+export default function useManagers(currentId: string) {
+  const [managers, setManagers] = useState<PersonModel[] | null>(null);
+
+  useMemo(() => {
+    getPeople().then((res) =>
+      setManagers(res.filter((p) => p.id !== currentId))
+    );
+  }, [currentId]);
+
+  return managers;
+}
```

Comment:
Maybe since we have a customHook to get people and a value "isManager" on our persons, we could use them like

```
import { useMemo } from 'react';
import usePeople from '../person/usePeople';

export default function useManagers(currentId: string) {
  const people = usePeople();
  return useMemo(() => people?.filter(person => person.isManager && person.id !== currentId), [people, currentId])
}

```

Reply:
It looks good to me but isn't it a bit "overkill" to call a custom hook that runs a useEffect inside of another custom hook that runs a useMemo ?

Reply:
Well the useMemo is definitely overkill anyway, I wouldn't put it in normal times unless absolutely necessary or because I see a performance improvement. I think the point is that when we're going to present this, we're going to show that this is possible and might give you a gain, but otherwise it's useless or sometimes could even hurt the performance since it's doing something on top.

In other ways, I think it's a great idea to put it in this "custom hook" which at this point is just a helper function if you really think about it. It allows it to be tested separately, keeps the logic away from the other components, and will allow to change one central point instead of multiple filter at different places (potentially different).

We just have to keep in mind than when presenting this, we need to point out that they need to test out in their app if this hinder or not their performance (if it's too small)

Reply:
Another solution could be to just put it in usePeople and deconstruct the manager, but I like it better this way at least. What's your thoughts on it ?

Reply:
I'm a little bit confused here, but my guess is we are using a `useMemo` that don't return anything, and make asynchronous call. So it should be a useEffect IMO.
BTW, it seems we are fetching the people just to filter the list (don't matter the way).
I would rather taking the list in parameter, then apply the filter cb, in a useMemo.
In this option, we need the list, obviously, so we can either : 

1. fetch the list in a useEffect / custom hook,
2. Use a context (with a custom hook ?)taht encapsulate every pages that need the list ; then the context is responsible for fetching the list (with a custom hook ? )

Option 2 is obviously more complex, but avoid unnecessary re-fetching. Please note in some cases refetching (option 1) would lead to fresher data, but also inconsistencies between components.

What do you thing about it ?

Reply:
> 2. Use a context (with a custom hook ?)taht encapsulate every pages that need the list ; then the context is responsible for fetching the list (with a custom hook ? )

Might be good because I think this is one very common use case where we have some data that will live through the app most of the time.

Maybe a `PeopleContext` around the components (well in our case, mostly everything) and use either a function passed by the context to get the manager specifically, and there a function to get the Manager or a customHook that consumes the context and filters (so similar to what I've done but instead of `usePeople()` we have a `useContext(PeopleContext).`

Reply:
Should we really implement a context for such a small use case? I changed the custom hook to this and it doesn't refetch anymore but updates the state directly. (Like @VictorJozwicki suggested)

`export default function useManagers(currentId: string) {
  const [managers, setManagers] = useState<PersonModel[] | null>(null);

  useEffect(() => {
    getPeople().then((res) =>
      setManagers(res.filter((p) => p.id !== currentId))
    );
  }, [currentId]);

  return managers;
}
`

Reply:
Context isn't "a big thing". Right now we would trigger an API call each time we go on an edit page.

Fetching every time we go to the edit page wouldn't be wrong per say in some cases. For example, if we know the data we want is changed/refreshed a lot server side. But we know that manager isn't a lot of people and doesn't change.

It would be interesting to say the different ways to keep the data fresh on our side, but we'll either say it orally or during the course explain things such as websocket or fetching in an interval.

We should implement a PeopleContext that englobes our App to avoid multiple API call.

Reply:
Please note that context not only avoid multi fetching, but also keep the whole app synchronized. 
For example, what if 

1. User fetch in order to display the list
2. Another user add a Person
3. User fetch the list again in order to edit and find managers

Depending on the UI/UX, User could have an issue (in this case seeing 5 people on the list, but 6 people as managers)
Using a "Single source of truth" is a very simple way to avoid that, and this is the reason why it's a very common best practice in ReactJS. Context is, IMO, a simple way to have a single source of truth.

That said, if refactoring this PR by using a Context seams a too large effort, we can let it as is, and create a ticket in the backlog. That's up to you guys :)

Reply:
We can always refactor after for sure

Reply:
Soooooo, should the context provide the custom hooks or should the custom hooks consume the context 🤔

Reply:
I guess it's possible to filter the managers and return it in the context directly, doing something like `const {managers} = useContext(PeopleContext)` but I'd rather use `const managers = useManager()`. 

It seem easier to use than knowing what's inside the context and which function to use.

So I vote to consume the context inside the customHook

Reply:
Yes I also think w should use a custom hook in order to abstract how value is consumed. This way, the caller just have to call the `usePepole` no matter when the data is stored.
Then, the `usePeople` cando whatever it wants, it's an implementation detail. 
I would recommend to create the context (`React.createContext` ) close to the hook, so that it's easier to understand that the hook need the context to work